hail/src/main/scala/is/hail/rvd/RVD.scala

```diff
@@ -1413,19 +1413,40 @@ object RVD {
       new RVD(typ, partitioner, crdd).checkKeyOrdering()
   }
 
+  private def copyFromType(destPType: PType, srcPType: PType, srcRegionValue: RegionValue): RegionValue =
+    RegionValue(srcRegionValue.region, destPType.copyFromType(srcRegionValue.region, srcPType, srcRegionValue.offset, false))
+
+  def unify(
+    rvds: Seq[RVD]
+  ): IndexedSeq[RVD] = rvds match {
+    case rvds.length == 1 => rvds.toIndexedSeq
+    case _ =>
+      if(rvds.forall(_.rowPType == rvds.head.rowPType)) {
+        return rvds.toIndexedSeq
+      }
+
+      val unifiedRowPType = InferPType.getNestedElementPTypesOfSameType(rvds.map(_.rowPType)).asInstanceOf[PStruct]
+
+      rvds.map(rvd => {
+        val newRVDType = rvd.typ.copy(rowType = unifiedRowPType)
+        rvd.map(newRVDType)(copyFromType(unifiedRowPType, rvd.rowPType, _))
+      })
+
   def union(
     rvds: Seq[RVD],
     joinKey: Int,
     ctx: ExecuteContext
   ): RVD = rvds match {
     case Seq(x) => x
     case first +: _ =>
-      assert(rvds.forall(_.rowPType == first.rowPType))
+      val newRVDs = unify(rvds)
```

Comment:
union doesn't need to change actually. Keep the assertion, it's on the user (TableUnion) to make sure the ptypes are the same

Reply:
sure thing